contracts/utils/Address.sol

```diff
@@ -59,30 +59,69 @@ library Address {
     }
 
     /**
-     * @dev Replacement for Solidity's low-level `call`: performs a low-level
-     * call with `data` to the target address `target`. Returns the `returndata`
-     * provided by the low-level call.
+     * @dev Performs a Solidity function call using a low level `call`. A
+     * plain`call` is an unsafe replacement for a function call: use this
+     * function instead.
      *
-     * The call is not executed if the target address is not a contract.
+     * If `target` reverts with a revert reason, it is bubbled up by this
+     * function (like regular Solidity function calls).
+     *
+     * Requirements:
+     *
+     * - `target` must be a contract.
+     * - calling `target` with `data` must not revert.
      */
     function functionCall(address target, bytes memory data) internal returns (bytes memory) {
       return functionCall(target, data, "Address: low-level call failed");
     }
 
     /**
-     * @dev Replacement for Solidity's low-level `call`: performs a low-level
-     * call with `data` to the target address `target`. Returns the `returndata`
-     * provided by the low-level call. Uses `errorMessage` as default revert message.
-     
-     * The call is not executed if the target address is not a contract.
+     * @dev Same as {Address-functionCall-address-bytes-}, but with
+     * `errorMessage` as a custom revert reason when `target` reverts.
      */
     function functionCall(address target, bytes memory data, string memory errorMessage) internal returns (bytes memory) {
+        return _functionCallWithValue(target, data, 0, errorMessage);
+    }
+
+    /**
+     * @dev Performs a Solidity function call using a low level `call`,
+     * transferring `value` wei. A plain`call` is an unsafe replacement for a
+     * function call: use this function instead.
+     *
+     * If `target` reverts with a revert reason, it is bubbled up by this
+     * function (like regular Solidity function calls).
+     *
+     * Requirements:
+     *
+     * - `target` must be a contract.
+     * - the calling contract must have an ETH balance of at least `value`.
+     * - calling `target` with `data` must not revert.
+     */
+    function functionCallWithValue(address target, bytes memory data, uint256 value) internal returns (bytes memory) {
+        return functionCallWithValue(target, data, value, "Address: low-level call with value failed");
+    }
+
+    /**
+     * @dev Same as {Address-functionCallWithValue-address-bytes-uint256-}, but
+     * with `errorMessage` as a custom revert reason when `target` reverts.
+     */
+    function functionCallWithValue(address target, bytes memory data, uint256 value, string memory errorMessage) internal returns (bytes memory) {
+        require(address(this).balance >= value, "Address: insufficient balance for call");
```

Comment:
This is interesting, does Solidity do this check too?

Reply:
No, the `call` reverts, so `success` is false, and you get the generic `low level call failed` revert reason. I added this check since that balance check is extremely cheap: https://eips.ethereum.org/EIPS/eip-1884#specification

Reply:
I wouldn't call that "extremely cheap"!! In fact, it's the same as the base cost of `CALL`, which should the dominating cost of `functionCall`. Adding this check increases that 2x. So I have a strong preference towards removing it.

Reply:
Hm, okay when transferring value the cost of a call increases by 9000 gas, so it would not be a 2x increase, but I'd still call it significant.

In any case, I think we should remove the check so as to be closer to Solidity semantics.

Reply:
`balance` costs 700 gas, but `selfbalance` (what that expression compiles into) costs 5 gas. That's what I was linking, `selfbalance` was introduced in that EIP. Sorry if this caused confusion.

Reply:
Ah, my bad!